### PR TITLE
Show all editable exercise attributes in detail pages

### PR DESCRIPTION
**Pull request by tessaRuckstuhl**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

It solves Issue #1434 (Show all exercise attributes in detail page)

### Description

I added following attributes to all exercise detail pages (text, modeling, programming and file upload)
  - categories
  - difficulty
  - assessmentDueDate
  - if mode team is set, the team sizes are shown 

Furthermore:

- if no dates, level or category is set, there is a corresponding text shown
- fixed link to course page in detail page
- Aswell, rearranged the order for all details pages:
  - all elements of the superclass, then elements of the subclass
  - related items together
  - same order as in create / edit

Note: For the programming detail page I followed a slightly different order, as it would not make sense to have the short title e.g. below the superclass elements but rather below the actual title
### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Management
3. Go to Exercises
4. Create an Exercise of each type and Save
5. View Exercise Detail Page and check for all attributes

### Screenshots

**Programming Exercise Screenshots**
image

**Modeling Exercise Screenshots**
image

**Text Exercise  Screenshots**

image

**File Upload Exercise Screenshots**
image

### CODE REVIEW

**Comment by krusche** [comment]
@tessaRuckstuhl thanks for creating this PR. Can you please post screenshots for all exercise detail pages. Please also double check that all common attributes of the super class exercise are displayed first (in the same order for each exercise type) and then the attributes of the sub classes are displayed.

**Comment by tessaRuckstuhl** [comment]
> @tessaRuckstuhl thanks for creating this PR. Can you please post screenshots for all exercise detail pages. Please also double check that all common attributes of the super class exercise are displayed first (in the same order for each exercise type) and then the attributes of the sub classes are displayed.

Sure! About the remark of showing all common attributes of the super class first: If I would restructure it like that, the order would not be the same as when creating/updating the exercises. 
Is this really wanted?

**Comment by krusche** [comment]
Yes this is fine. Please try to keep related items together (eg dates) and if possible try to use the same order as in the create/edit dialog

**Comment by tessaRuckstuhl** [comment]
> Yes this is fine. Please try to keep related items together (eg dates) and if possible try to use the same order as in the create/edit dialog

Ok, so the current order should be fine then.

**Comment by tessaRuckstuhl** [comment]
So to get this clear - using the order of the create/edit dialog + grouping related items is fine? Not necessarily super/sub class ordering?

**Comment by krusche** [comment]
I would suggest to order after the following priorities:
1) all elements of the superclass, then elements of the subclass
2) related items together
3) same order as in create / edit

1) is highest priority and should be fulfilled, but leaves some space for ordering the elements of the super class. Then 2) should be followed...

**Comment by tessaRuckstuhl** [comment]
Okay got it! Thanks.

**Comment by tessaRuckstuhl** [comment]
I found a error in the programming exercise detail page when (re-)loading. E.g. https://artemistest.ase.in.tum.de/#/course-management/226/programming-exercises/1993
I'm not sure now whether I should deploy it to the same testserver, as it might seem like the errors origin from this PR. Should I just ignore that and deploy anyways? 
Aswell for future reference: Where should I report errors found in the latest version of the development branch and/or testserver? 
image

**Comment by krusche** [comment]
> I found an error in the programming exercise detail page when (re-)loading on testserver1. The "**internal server error**" is shown three times, but then disappears aftar a few seconds. E.g. https://artemistest.ase.in.tum.de/#/course-management/226/programming-exercises/1993
> I'm not sure now whether I should deploy this PR to the same testserver, as it might seem like the errors origin from this PR. Should I just ignore that and deploy anyways?
> Aswell for future reference: Where should I report errors found in the latest version of the development branch and/or testserver?
> image

Please create an issue on GitHub and report it to me. This issue should be already fixed in the latest version on develop

**Comment by krusche** [comment]
There are some merge conflicts, which should hopefully be easy to fix.
@tessaRuckstuhl could you please fix them?
I think you can just use the changes from this branch and re-apply the date pipe changes in https://github.com/ls1intum/Artemis/pull/1488 by @sascha11110

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html [hunk 1]
```diff
@@ -35,6 +35,12 @@ <h2><span jhiTranslate="artemisApp.programmingExercise.detail.title">Programming
                 </button>
             </div>
             <dl class="row-md jh-entity-details">
+                <dt><span jhiTranslate="artemisApp.exercise.course">Course</span></dt>
```
Comment (jpbernius) [inline src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html:38]: May I ask what is preventing you from using your `ExerciseDetailsComponent` here as well?

File: src/main/webapp/app/exercises/shared/exercise/exercise-details/exercise-details.component.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    selector: 'jhi-exercise-details',
+    templateUrl: './exercise-details.component.html',
+    styleUrls: ['./exercise-details.component.scss'],
+})
+export class ExerciseDetailsComponent implements OnInit {
+    @Input() exerciseDetails: Exercise;
+
+    AssessmentType = AssessmentType;
+
+    superExerciseDetails: Exercise;
```
[the remark this answers is not in the document]
Comment (tessaRuckstuhl) [reply src/main/webapp/app/exercises/shared/exercise/exercise-details/exercise-details.component.ts:18]: We don't. Thanks for noticing, not sure why I did this anymore.

File: src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html (hunk 1 above)
Comment (tessaRuckstuhl) [reply src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html:38]: A few things: The programming exercise details has a few attributes which are either missing in the super exercise details (e.g. `buildAndTestStudentSubmissionsAfterDueDate`) and as well has a short name, which should imo directly follow the title. As well the problem statement and grading instructions are formatted differently.

Comment (tessaRuckstuhl) [reply src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html]: Just added a commit, which tries to reuse as much of the component in that exercise detail page, too. Let me know what you think.

**Comment by krusche** [comment]
One small additional change request. Can you please remove the 2 "optional" in the programming exercise page? For solution repo url and test repo url

**Pull request merged**